### ISSUE TITLE
The computation of loss of AspectExtractor

### ISSUE THREAD

**Issue by Yuran-Zhao** [sfu-natlang/SFUTranslate#36]
Hi! I notice that the code of computing the loss of AspectExtrator [here](https://github.com/sfu-natlang/SFUTranslate/blob/0b3762e93bdc446e681f7f82a13455d580d26827/translate/models/aspects/module.py#L118) is inconsistent with what you described in your paper "Better Neural Machine Translation by Extracting Linguistic Information from BERT".

It seems should be `loss += cnt - lss / ((len(self.encoders)-1)*(len(self.encoders)-2)/2)` or something like that.

I wonder if I missed something? It makes me a little confused.

**Comment by shavarani** [comment]
I'm sorry I was away for a while, I think the line that you highlighted is adding the average distance of the produced vectors to the loss. The only difference with the paper is that in Equation 4, in the paper, we calculate this value as one minus the average similarity of the vectors. The implementation modifies it to speed up the calculation.

**Comment by Yuran-Zhao** [comment]
Thanks for your reply! 
Equation 4 in your paper is understandable for me. If two aspect vectors are similar to each other, the total loss will be higher. Nevertheless, the implementation just **adds** the similarity of the vectors directly into the loss, which means the model needs to generate similar vectors to make the loss decrease.
So, maybe it should simply be `loss -= lss / cnt` ?

**Comment by shavarani** [comment]
Thanks for pointing this out, looks like we have had a re-implementation error when releasing the code for the paper. It is correct the loss needs to be negated.
